### PR TITLE
GH-44555: [C++][Compute] Allow casting struct to bigger nullable struct

### PR DESCRIPTION
**Pull request by Tom-Newton**
### Rationale for this change
Sometimes its useful to add a column full of nulls in a cast. Currently this is supported in top level columns but not inside structs. Example where this is important: https://github.com/delta-io/delta-rs/issues/1610

### What changes are included in this PR?
Add support for filling in columns full of null for nullable struct fields. 
I've gone for a fairly minimal change that achieves what I needed but I wonder if there should be a more significant change so that this casting is done by field name and ignore the field order. 

### Are these changes tested?
Yes. The expected behaviour in several existing tests has been altered and a couple of new tests have been added.

I also rolled out a custom build with this change internally because it suddenly became a critical problem. 

### Are there any user-facing changes?
Yes. There are scenarios that previously failed with `struct fields don't match or are in the wrong order` but now succeed after filling in `null`s.

* GitHub Issue: #44555

### CODE REVIEW

**Comment by Tom-Newton** [comment]
I'm not confident that I've taken the right approach here but I think its ready for review.

**Comment by Tom-Newton** [comment]
Ok, it looks like there is at least one more tests I need to update. I can probably sort that tomorrow.

File: cpp/src/arrow/dataset/scanner_test.cc [hunk 1]
```diff
@@ -2328,7 +2328,7 @@ DatasetAndBatches MakeNestedDataset() {
       field("b", boolean()),
       field("c", struct_({
                      field("d", int64()),
-                     field("e", float64()),
+                     field("e", int64()),
```
Comment (Tom-Newton) [inline cpp/src/arrow/dataset/scanner_test.cc:2331]: I suspect this was not supposed to be different from in the physical schema. This data is used to test if a virtual column will be materialised so I guess it shouldn't be testing int64 -> float64 casting at the same time.

**Comment by Tom-Newton** [comment]
Sorry for the direct ping but @lidavidm please could you review when you get a chance. I assume you are the right person to review this?

File: cpp/src/arrow/compute/kernels/scalar_cast_nested.cc [hunk 2]
```diff
@@ -348,25 +349,47 @@ struct CastStruct {
...
-      return Status::TypeError(
-          "struct fields don't match or are in the wrong order: Input fields: ",
-          in_type.ToString(), " output fields: ", out_type.ToString());
+      if (all_in_field_names.count(out_field->name()) == 0 && out_field->nullable()) {
+        // Didn't match current in_field, but we can fill with null.
+        // Filling with null is only acceptable on nuallable fields when there
+        // is definitely no in_field with matching name.
+
+        // -2 is a sentinel value indicating fill with null.
+        fields_to_select[out_field_index++] = -2;
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_cast_nested.cc:381]: nit: can we add something like `constexpr int kFillNullSentinel = -2`?

File: cpp/src/arrow/compute/kernels/scalar_cast_nested.cc [hunk 3]
```diff
@@ -348,25 +349,47 @@ struct CastStruct {
...
       }
-    }
-
-    if (out_field_index < out_field_count) {
-      return Status::TypeError(
-          "struct fields don't match or are in the wrong order: Input fields: ",
-          in_type.ToString(), " output fields: ", out_type.ToString());
+      if (all_in_field_names.count(out_field->name()) == 0 && out_field->nullable()) {
+        // Didn't match current in_field, but we can fill with null.
+        // Filling with null is only acceptable on nuallable fields when there
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_cast_nested.cc:377]:
```suggestion
        // Filling with null is only acceptable on nullable fields when there
```

**Review by lidavidm — COMMENTED** [review]
Looks reasonable to me, thanks. CC @bkietz for some more eyes?

File: cpp/src/arrow/compute/kernels/scalar_cast_nested.cc (hunk 2 above)
Comment (Tom-Newton) [reply cpp/src/arrow/compute/kernels/scalar_cast_nested.cc]: Done but I wasn't really confident on the best place to define the constant.

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_cast_nested.cc]: It can be placed inside the method definition itself!

Comment (Tom-Newton) [reply cpp/src/arrow/compute/kernels/scalar_cast_nested.cc]: Thanks, I moved it inside. As you can probably tell I am a C++ noob.

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Ok, I don't think that one CI failure is related here. Merging now. Thanks @Tom-Newton, sorry for the delays!

**Comment by Tom-Newton** [comment]
> Ok, I don't think that one CI failure is related here. Merging now. Thanks @Tom-Newton, sorry for the delays!

Thanks for reviewing

**Pull request merged**